### PR TITLE
codegen "unreachable" for invalid SetDiscriminant

### PR DESCRIPTION
**Pull request by RalfJung**
Follow-up from https://github.com/rust-lang/rust/pull/66960. I also realized I don't understand our policy for using `abort` vs `unreachable`. AFAIK `abort` is safe to call and just aborts the process, while `unreachable` is UB. But sometimes we use both, like here

https://github.com/rust-lang/rust/blob/d825e35ee8325146e6c175a4c61bcb645b347d5e/src/librustc_codegen_ssa/mir/block.rs#L827-L828

and here

https://github.com/rust-lang/rust/blob/d825e35ee8325146e6c175a4c61bcb645b347d5e/src/librustc_codegen_ssa/mir/block.rs#L264-L265

The second case is even more confusing because that looks like an unreachable `return` to me, so why would we codegen a safe abort there?

r? @eddyb Cc @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
Well that is curious:
```
error: failed to parse bitcode for LTO module: Invalid instruction with no BB (Producer: 'LLVM9.0.0-rust-1.41.0-dev-de1a7dbf6' Reader: 'LLVM 9.0.0-rust-1.41.0-dev-de1a7dbf6')
```

**Comment by eddyb** [comment]
> The second case is even more confusing because that looks like an unreachable return to me, so why would we codegen a safe abort there?

LLVM `call`s still require the rest of the block after them, even if the callee never returns.

File: src/librustc_codegen_ssa/mir/place.rs [hunk 1]
```diff
@@ -333,6 +333,7 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
         variant_index: VariantIdx
     ) {
         if self.layout.for_variant(bx.cx(), variant_index).abi.is_uninhabited() {
+            bx.unreachable();
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/place.rs]: You probably need to start a new block, or stop the remaining MIR statements/terminator from being codegen'd.

cc @rkruppe @nagisa

**Comment by RalfJung** [comment]
> LLVM calls still require the rest of the block after them, even if the callee never returns.

That explains the first case (the `Abort` terminator). But why does an unreachable return get translated to a *well-defined* abort (as opposed to an immediate-UB "unreachable")?

File: src/librustc_codegen_ssa/mir/place.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_codegen_ssa/mir/place.rs]: For now I made this an `abort`, also matching what we do for uninhabited returns. But it seems odd to me that we'd give those well-defined "trap" semantics as opposed to exploiting their UB.

File: src/librustc_codegen_ssa/mir/block.rs [hunk 2]
```diff
@@ -261,7 +261,10 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         if self.fn_abi.ret.layout.abi.is_uninhabited() {
             // Functions with uninhabited return values are marked `noreturn`,
             // so we should make sure that we never actually do.
+            // We play it safe by using a well-defined `abort`, and not immediate UB.
             bx.abort();
```
Comment (RalfJung) [inline src/librustc_codegen_ssa/mir/block.rs]: I am documenting existing behavior here, but my own inclination would be to remove the `abort` here to give the optimizer some more information. Or should we wait until someone determines that this is actually costing us some optimizations?

**Comment by oli-obk** [comment]
One difference between the two is that `unreachable` is an instruction (https://llvm.org/docs/LangRef.html#unreachable-instruction) while `trap`/`abort` is an intrinsic (https://llvm.org/docs/LangRef.html#llvm-trap-intrinsic).

I don't know what difference that makes though. Technically it should suffice to just emit unreachables.

**Comment by RalfJung** [comment]
The problem is that emitting `unreachable` followed by other stuff in the same basic block leads to ICEs.

**Comment by oli-obk** [comment]
Maybe emitting `unreachable` should "poison" the builder so any further commands trying to add things to the current block will just not do anything. The poisoning can be cleared when a new block is started.

Alternatively we could add some scheme where an `Err(Unreachable)` is bubbled up to the root code that generated it and then that code is supposed to stop generating the current block

**Comment by RalfJung** [comment]
The builder has a notion of "current block", right? Maybe after calling `unreachable`  we should make a new block (with no incoming edge) and make that the current block.

But I don't know enough about our codegen APIs to implement any of these proposals.

**Comment by oli-obk** [comment]
Let's get this merged and I'll push making all of the traps `unreachable` onto my stack of PRs to make

@bors r+

**Comment by RalfJung** [comment]
Not all of them I hope. ;) The `Abort` MIR terminator should remain a well-defined trap.

**Comment by oli-obk** [comment]
:joy: right, only those that are unreachable at runtime

**Comment by eddyb** [comment]
Hang on, none of us know why the aborts are there!
Please don't remove any in the future without first talking to some codegen people who might know, or by finding a past decision by digging in the git history.

> The problem is that emitting `unreachable` followed by other stuff in the same basic block leads to ICEs.

I didn't check but maybe LLVM produces an assertion if they are enabled (do you have them on?).
But the important thing is that this is a problem with *all* LLVM block terminators, LLVM will let you add more instructions to the block after the terminator, but it's malformed if you do.

As for "poisoning the builder"... Yeah that massive complication which we had for a very long time and which was made unnecessary by MIR :rofl:

**Comment by oli-obk** [comment]
> Yeah that massive complication which we had for a very long time and which was made unnecessary by MIR rofl

We can also monomorphize the MIR, rerun optimizations and have some guaranteed optimizations that cause no such weird code to even be given to llvm. The monorphized MIR can then be thrown away right after codegen. No need to keep it around

**Comment by RalfJung** [comment]
> Please don't remove any in the future without first talking to some codegen people who might know, or by finding a past decision by digging in the git history.

Fair.

OTOH, given there is no comment in the code, if there was anything subtle going on then someone seriously screwed up by not documenting it right...

**Pull request merged**